File: Kotlin/FourNumbersSum.kt
package com.algoexpert.program

fun fourNumberSum(array: MutableList<Int>, targetSum: Int): List<List<Int>> {
    val allPairSums = mutableMapOf<Int, MutableList<MutableList<Int>>>()
	val quadruplets = mutableListOf<MutableList<Int>>()
	for (i in 1 until array.size - 1) {
		for(j in i + 1 until array.size) {
			val currentSum = array[i] + array[j]
			val difference = targetSum - currentSum
			if (allPairSums.containsKey(difference)) {
				for (pair in allPairSums[difference]!!) {
					val p = pair.toMutableList<Int>()
					p.add(array[i])
					p.add(array[j])
					quadruplets.add(p)
				}
			}
		}
		for (k in 0 until i) {
			val currentSum = array[i] + array[k]
			if(!allPairSums.containsKey(currentSum)) {
				allPairSums[currentSum] = mutableListOf<MutableList<Int>>()
				allPairSums[currentSum]!!.add(mutableListOf<Int>(array[k], array[i]))
			} else {
				allPairSums[currentSum]!!.add(mutableListOf<Int>(array[k], array[i]))
			}
		} 
	}
    return quadruplets
}

File: C++/FourNumbersSum.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>
using namespace std;

vector<vector<int>> fourNumberSum(vector<int> array, int targetSum) {
  unordered_map<int, vector<vector<int>>> allPairSums;
	vector<vector<int>> quadruplets{};
	for(int i = 1; i < array.size() - 1; i++) {
		for (int j = i+1; j < array.size(); j++) {
			int currentSum = array[i] + array[j];
			int difference = targetSum - currentSum;
			if(allPairSums.find(difference) != allPairSums.end()) {
				for (vector<int> pair : allPairSums[difference]) {
					pair.push_back(array[i]);
					pair.push_back(array[j]);
					quadruplets.push_back(pair);
				}
			}
		}
		for(int k = 0; k < i; k++) {
			int currentSum = array[i] + array[k];
			if(allPairSums.find(currentSum) == allPairSums.end()){
				allPairSums[currentSum] = vector<vector<int>>{{array[k], array[i]}};
			} else {
				allPairSums[currentSum].push_back(vector<int>{array[k], array[i]});
			}
		}
	}
  return quadruplets;
}